calico_node/startup/startup.go

```diff
@@ -781,83 +855,154 @@ func ensureDefaultConfig(cfg *api.CalicoAPIConfig, c *client.Client, node *api.N
 	// for containers to speak to services running on the host (e.g. k8s
 	// pods speaking to k8s api-server, and mesos tasks registering with agent
 	// on startup).  Note: KDD does not yet support per-node felix config.
-	if cfg.Spec.DatastoreType != api.Kubernetes {
-		if err := ensureNodeFelixConfig(c, node.Metadata.Name, "DefaultEndpointToHostAction", "RETURN"); err != nil {
-			return err
+	if cfg.Spec.DatastoreType != apiconfig.Kubernetes {
+		felixNodeCfg, err := c.FelixConfigurations().Get(ctx, fmt.Sprintf("%s%s", FELIX_NODE_CONFIG_NAME_PREFIX, node.Name), options.GetOptions{})
+		if err != nil {
+			// Create the default config if it doesn't already exist.
+			if _, ok := err.(cerrors.ErrorResourceDoesNotExist); ok {
+				newFelixNodeCfg := api.NewFelixConfiguration()
+				newFelixNodeCfg.Name = fmt.Sprintf("%s%s", FELIX_NODE_CONFIG_NAME_PREFIX, node.Name)
+				newFelixNodeCfg.Spec.DefaultEndpointToHostAction = "RETURN"
+				_, err = c.FelixConfigurations().Create(ctx, newFelixNodeCfg, options.SetOptions{})
+				if err != nil {
+					log.WithError(err).WithField("FelixConfig", newFelixNodeCfg).Errorf("Error creating Felix node config")
+					return err
+				}
+			} else {
+				log.WithError(err).WithField("FelixConfig", FELIX_NODE_CONFIG_NAME_PREFIX).Errorf("Error getting Felix node config")
+				return err
+			}
+		} else {
+			if felixNodeCfg.Spec.DefaultEndpointToHostAction == "" {
+				felixNodeCfg.Spec.DefaultEndpointToHostAction = "RETURN"
+				_, err = c.FelixConfigurations().Update(ctx, felixNodeCfg, options.SetOptions{})
+				if err = checkConflictError(err); err != nil {
+					log.WithError(err).WithField("FelixConfig", felixNodeCfg).Errorf("Error updating Felix node config")
+					return err
+				}
+			} else {
+				log.WithField("DefaultEndpointToHostAction", felixNodeCfg.Spec.DefaultEndpointToHostAction).Debug("Host Felix value already assigned")
+			}
 		}
 	}
 
 	// Store the Calico Version as a global felix config setting.
-	if err := checkConflictError(c.Config().SetFelixConfig("CalicoVersion", "", VERSION)); err != nil {
-		return err
+	clusterInfo, err := c.ClusterInformation().Get(ctx, GLOBAL_CLUSTER_INFO_NAME, options.GetOptions{})
+	if err != nil {
```

Comment:
There are three things we update in the ClusterInformation:  the GUID, the cluster type and the version.   We should do this as part of the same operation otherwise we end up doing more queries/updates than we need to.